File: extension/src/psl.ts
// Registrable-domain (eTLD+1) resolution against the vendored PSL snapshot — TS mirror of
// core/.../client/autofill/Psl.kt (design docs/design/2026-07-10-etld1-psl-matching.md,
// spec 02 §3.1 amendment). All three impls run the SAME spec/test-vectors/urimatch-etld1.json.
//
// BUNDLE PLACEMENT (A8): this module is the ONLY importer of pslData (~144 KB). urimatch.ts
// takes the resolver as a parameter precisely so content-script bundles never pull the blob —
// keep it that way.

import { PSL_RULES_JOINED, PSL_SNAPSHOT_HASH } from "./pslData.ts"; // explicit .ts: node --test type-stripping needs it (web mirror stays extensionless)
import type { PslResolve, PslResult } from "./urimatch.ts";

export { PSL_SNAPSHOT_HASH };

let parsed: { exact: Set<string>; wildcards: Set<string>; exceptions: Set<string> } | null = null;

/** Rule sets parsed once per process. Wildcards stored as their base ("*.kawasaki.jp" →
 *  "kawasaki.jp"), exceptions minus the "!". */
function sets(): NonNullable<typeof parsed> {
  if (!parsed) {
    const exact = new Set<string>();
    const wildcards = new Set<string>();
    const exceptions = new Set<string>();
    for (const rule of PSL_RULES_JOINED.split("\n")) {
      if (!rule) continue;
      if (rule.startsWith("!")) exceptions.add(rule.slice(1));
      else if (rule.startsWith("*.")) wildcards.add(rule.slice(2));
      else exact.add(rule);
    }
    parsed = { exact, wildcards, exceptions };
  }
  return parsed;
}

/**
 * Resolve a NORMALIZED host (the caller runs normalizeHost first; this only self-guards).
 *
 * Three-state result (A1): "registrable" (positively known), "public-suffix" (the host IS a
 * public suffix — exact, wildcard-derived, or exception-derived alike), "unknown" (no
 * EXPLICIT rule matched — the PSL's implicit `*` fallback is deliberately NOT applied — or
 * non-ASCII / IP / unparseable input; matching degrades to the pre-eTLD+1 rules).
 *
 * Walk (A2, normative): candidate suffixes longest-first; per candidate test exception →
 * exact → wildcard; an exception sets suffixLen = candidateLabels − 1 and STOPS (never
 * overridden by its sibling wildcard); exact/wildcard set suffixLen = candidateLabels and
 * stop. Longest-first + stop-on-first makes the longest rule win by construction.
 */
export const pslResolve: PslResolve = (host: string): PslResult => {
  // eslint-disable-next-line no-control-regex
  if (!host || /[^\x00-\x7f]/.test(host) || host.includes(":")) return { kind: "unknown" };
  const labels = host.split(".");
  if (labels.some((l) => !l)) return { kind: "unknown" };
  if (labels.length === 4 && labels.every((l) => /^\d+$/.test(l))) return { kind: "unknown" }; // IPv4
  const { exact, wildcards, exceptions } = sets();
  let suffixLen = 0;
  for (let i = 0; i < labels.length; i++) {
    const cand = labels.slice(i).join(".");
    if (exceptions.has(cand)) {
      suffixLen = labels.length - i - 1;
      break;
    }
    if (exact.has(cand)) {
      suffixLen = labels.length - i;
      break;
    }
    if (labels.length - i >= 2 && wildcards.has(labels.slice(i + 1).join("."))) {
      suffixLen = labels.length - i;
      break;
    }
  }
  if (suffixLen === 0) return { kind: "unknown" };
  if (suffixLen >= labels.length) return { kind: "public-suffix" };
  return { kind: "registrable", domain: labels.slice(labels.length - suffixLen - 1).join(".") };
};

File: core/src/commonMain/kotlin/io/silencelen/andvari/core/client/autofill/Psl.kt
package io.silencelen.andvari.core.client.autofill

/**
 * Registrable-domain (eTLD+1) resolution against the vendored PSL snapshot
 * (design docs/design/2026-07-10-etld1-psl-matching.md, spec 02 §3.1).
 *
 * Three-state result (amendment A1 — a nullable String cannot express the bare-suffix
 * tightening, and exact-rule membership misses all wildcard-derived suffixes):
 *  - [Registrable]  — the host's registrable domain is positively known.
 *  - [PublicSuffix] — the host IS a public suffix (exact, wildcard-derived, or
 *                     exception-derived alike): `github.io`, `co.uk`, `b.kawasaki.jp`,
 *                     `us-east-1.compute.amazonaws.com`.
 *  - [Unknown]      — no EXPLICIT rule matched (the PSL's implicit `*` fallback is
 *                     deliberately NOT applied — fail-safe A3/R-OLD), or the input is
 *                     non-ASCII / an IP literal / unparseable. Matching degrades to the
 *                     pre-eTLD+1 rules for Unknown, so `nas.local`-style intranet hosts
 *                     keep today's behavior bit-for-bit.
 */
sealed interface PslResult {
    data class Registrable(val domain: String) : PslResult
    data object PublicSuffix : PslResult
    data object Unknown : PslResult
}

object Psl {
    // Rule sets parsed once from the generated snapshot. Wildcard rules are stored as their
    // base (`*.kawasaki.jp` → `kawasaki.jp`), exceptions minus the `!`.
    private val exact: Set<String> by lazy { sets.first }
    private val wildcards: Set<String> by lazy { sets.second }
    private val exceptions: Set<String> by lazy { sets.third }

    private val sets: Triple<Set<String>, Set<String>, Set<String>> by lazy {
        val e = HashSet<String>()
        val w = HashSet<String>()
        val x = HashSet<String>()
        for (rule in PslData.joined.splitToSequence('\n')) {
            when {
                rule.isEmpty() -> {}
                rule.startsWith("!") -> x.add(rule.substring(1))
                rule.startsWith("*.") -> w.add(rule.substring(2))
                else -> e.add(rule)
            }
        }
        Triple(e, w, x)
    }

    /**
     * Resolve a NORMALIZED host (lowercase, no port/userinfo/trailing dot — the caller runs
     * [UriMatch.normalizeHost] first; this only self-guards).
     *
     * Walk (amendment A2, normative): candidate suffixes longest-first; at each candidate
     * test exception → exact → wildcard; an exception sets `suffixLen = candidateLabels − 1`
     * and STOPS (never overridden by the sibling wildcard that always co-exists); exact or
     * wildcard sets `suffixLen = candidateLabels` and stops. Longest-first + stop-on-first
     * makes the longest rule win by construction.
     */
    fun resolve(host: String): PslResult {
        if (host.isEmpty() || host.any { it.code >= 128 } || host.contains(':')) return PslResult.Unknown
        val labels = host.split('.')
        if (labels.any { it.isEmpty() }) return PslResult.Unknown
        if (labels.size == 4 && labels.all { l -> l.all { it.isDigit() } }) return PslResult.Unknown // IPv4
        var suffixLen = 0
        for (i in labels.indices) {
            val cand = labels.subList(i, labels.size).joinToString(".")
            if (cand in exceptions) {
                suffixLen = (labels.size - i) - 1
                break
            }
            if (cand in exact) {
                suffixLen = labels.size - i
                break
            }
            if (labels.size - i >= 2 && labels.subList(i + 1, labels.size).joinToString(".") in wildcards) {
                suffixLen = labels.size - i
                break
            }
        }
        return when {
            suffixLen == 0 -> PslResult.Unknown
            suffixLen >= labels.size -> PslResult.PublicSuffix
            else -> PslResult.Registrable(labels.subList(labels.size - suffixLen - 1, labels.size).joinToString("."))
        }
    }
}
